### ISSUE TITLE
Segfault if IPv4 is not available

### ISSUE THREAD

**Issue by whyman** [gerbera/gerbera#1090]
I have issues with Gerbera bootstrap on Gentoo Linux. 
Gerbera will start successfully after boot, once all the interfaces are on-line, etc.

During the boot process, Gentoo, or at least my system, brings IPv6 on-line before IPv4, with IPv4 arriving about 3 seconds later. 

As such, the startup script, which says Gerbera needs the net, thinks the net is on-line due to IPv6 availability and happily starts up Gerbera, which promptly crashes because IPv4 is not yet up.

Whilst not exactly the problem as described initially in this issue, I figure that should the connect to the IPv4 socket fail, maybe Gerbera could wait and retry, or, fail gracefully. The startup script on Gentoo can be configured to handle graceful failures and is happy with delays in startup but it gets messy when the app fails with a segfault.

Syslog of gerbera startup crash follows (lightly edited to remove superfluous entries and to fluff IPv6 addresses) :-

```
Jan  2 22:03:02 Lyalls-PC kernel: [   19.625318] IPv6: ADDRCONF(NETDEV_CHANGE): enp0s31f6: link becomes ready
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: DUID 00:01:00:01:21:61:12:f7:70:4d:7b:67:83:60
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: IAID 7b:67:83:60
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding address fe80::16:cb74:9a0b:????
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: rebinding lease of 192.168.1.9
Jan  2 22:03:02 Lyalls-PC dhcpcd[1897]: enp0s31f6: probing address 192.168.1.9/24
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: soliciting an IPv6 router
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: Router Advertisement from fe80::be05:43ff:fee6:????
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding address 2001:44b8:24f:6200:4f72:303e:????:15dd/64
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding route to 2001:44b8:24f:????::/64
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: requesting DHCPv6 information
Jan  2 22:03:03 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding default route via fe80::be05:43ff:fee6:????
Jan  2 22:03:06 Lyalls-PC kernel: [   23.133990] gerbera[2424]: segfault at 40 ip 000055d3be79414f sp 00007ffed7c5ffc0 error 4 in gerbera[55d3be6cc000+1bf000]
Jan  2 22:03:06 Lyalls-PC kernel: [   23.133993] Code: 66 ee 14 00 00 0f 84 e0 03 00 00 b8 ff ff ff ff f0 41 0f c1 44 24 08 83 f8 01 0f 84 e1 03 00 00 0f 1f 00 48 8b b5 70 01 00 00 <48> 8b 46 40 48 39 46 38 0f 84 73 01 00 00 31 db 4c 8d 7c 24 20 4c
Jan  2 22:03:06 Lyalls-PC /etc/init.d/gerbera[2423]: start-stop-daemon: /usr/bin/gerbera died
Jan  2 22:03:06 Lyalls-PC /etc/init.d/gerbera[2402]: ERROR: gerbera failed to start
Jan  2 22:03:07 Lyalls-PC ntpd[2266]: Listen normally on 6 enp0s31f6 [2001:44b8:24f:6200:4f72:303e:5f61:????]:123
Jan  2 22:03:08 Lyalls-PC dhcpcd[1897]: enp0s31f6: leased 192.168.1.9 for 864000 seconds
Jan  2 22:03:08 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding route to 192.168.1.0/24
Jan  2 22:03:08 Lyalls-PC dhcpcd[1897]: enp0s31f6: adding default route via 192.168.1.254

```

_Originally posted by @lyallp in https://github.com/gerbera/gerbera/issues/903#issuecomment-754339558_

**Comment by lyallp** [comment]
I have the kernel supposedly dumping cores in **/data/cores/core.%e.%p.%u.%s.%t** according to **/proc/sys/kernel/core_pattern**
but the directory is empty. :(

I also have a cron job that cleans that directory of cores older than 7 days, so that does not explain why there is no core as there has not been 7 days since the last crash.

I will set 'ulimit 1000000' in the gerbera startup script to see if I can capture a core and report back.

I speculate that, in order to reproduce the problem, disable IPv4 whilst leaving IPv6 available then run Gerbera.

**Comment by lyallp** [comment]
I have the kernel supposedly dumping cores in /data/cores/core.%e.%p.%u.%s.%t according to **/proc/sys/kernel/core_pattern**
but the directory is empty. :(

I also have a cron job that cleans that directory of cores older than 7 days, so that does not explain why there is no core as there has not been 7 days since the last crash.

I tried setting 'ulimit 1000000' in the gerbera startup script to see if I can capture a core and report back, but no core is generated. It's not until 3 seconds later does the kernel report that it is placing cores in /data/cores and not being hugely familiar with internals of the kernal, I am unsure as to if/where it places cores at this stage of the boot.

I speculate that, in order to reproduce the problem, disable IPv4 whilst leaving IPv6 available then run Gerbera.

Gerbera log below. It would appear the problem is handling an error generated by libupnp, not the fact that libupnp actually generated an error, which is another issue :)

```
2021-01-03 09:00:01   info: Gerbera UPnP Server version 1.6.1 - http://gerbera.io/
2021-01-03 09:00:01   info: ===============================================================================
2021-01-03 09:00:01   info: Gerbera is free software, covered by the GNU General Public License version 2
2021-01-03 09:00:01   info: Copyright 2016-2020 Gerbera Contributors.
2021-01-03 09:00:01   info: Gerbera is based on MediaTomb: Copyright 2005-2010 Gena Batsyan, Sergey Bostandzhyan, Leonhard Wimmer.
2021-01-03 09:00:01   info: ===============================================================================
2021-01-03 09:00:01   info: Loading configuration from: /etc/gerbera/config.xml
2021-01-03 09:00:01   info: Checking configuration...
2021-01-03 09:00:01   info: Setting filesystem import charset to ANSI_X3.4-1968
2021-01-03 09:00:01   info: Setting metadata import charset to ANSI_X3.4-1968
2021-01-03 09:00:01   info: Setting playlist charset to ANSI_X3.4-1968
2021-01-03 09:00:01   info: Configuration check succeeded.
2021-01-03 09:00:03   info: About to migrate metadata from mt_cds_object to mt_metadata
2021-01-03 09:00:03   info: No data will be removed from mt_cds_object
2021-01-03 09:00:04   info: Migrated metadata - object count: 0
2021-01-03 09:00:04  error: LibUPnP error code: -121
2021-01-03 09:00:04  error: UpnpUnRegisterClient failed (-116)
2021-01-03 09:00:04  error: UpnpUnRegisterRootDevice failed (-116)
```

**Comment by lyallp** [comment]
I have the kernel supposedly dumping cores in /data/cores/core.%e.%p.%u.%s.%t according to **/proc/sys/kernel/core_pattern**
but the directory is empty. :(

I also have a cron job that cleans that directory of cores older than 7 days, so that does not explain why there is no core as there has not been 7 days since the last crash.

I tried setting 'ulimit 1000000' in the gerbera startup script to see if I can capture a core and report back, but no core is generated. It's not until 3 seconds later does the kernel report that it is placing cores in /data/cores and not being hugely familiar with internals of the kernal, I am unsure as to if/where it places cores at this stage of the boot.

I speculate that, in order to reproduce the problem, disable IPv4 whilst leaving IPv6 available then run Gerbera.

Gerbera log below. It would appear the problem is handling an error generated by libupnp, not the fact that libupnp actually generated an error, which is another issue :)

```
2021-01-03 09:00:01   info: Gerbera UPnP Server version 1.6.1 - http://gerbera.io/
2021-01-03 09:00:01   info: ===============================================================================
2021-01-03 09:00:01   info: Gerbera is free software, covered by the GNU General Public License version 2
2021-01-03 09:00:01   info: Copyright 2016-2020 Gerbera Contributors.
2021-01-03 09:00:01   info: Gerbera is based on MediaTomb: Copyright 2005-2010 Gena Batsyan, Sergey Bostandzhyan, Leonhard Wimmer.
2021-01-03 09:00:01   info: ===============================================================================
2021-01-03 09:00:01   info: Loading configuration from: /etc/gerbera/config.xml
2021-01-03 09:00:01   info: Checking configuration...
2021-01-03 09:00:01   info: Setting filesystem import charset to ANSI_X3.4-1968
2021-01-03 09:00:01   info: Setting metadata import charset to ANSI_X3.4-1968
2021-01-03 09:00:01   info: Setting playlist charset to ANSI_X3.4-1968
2021-01-03 09:00:01   info: Configuration check succeeded.
2021-01-03 09:00:03   info: About to migrate metadata from mt_cds_object to mt_metadata
2021-01-03 09:00:03   info: No data will be removed from mt_cds_object
2021-01-03 09:00:04   info: Migrated metadata - object count: 0
2021-01-03 09:00:04  error: LibUPnP error code: -121
2021-01-03 09:00:04  error: UpnpUnRegisterClient failed (-116)
2021-01-03 09:00:04  error: UpnpUnRegisterRootDevice failed (-116)
```

**Comment by whyman** [comment]
If we get an error when initialising libupnp, we simply log it and then exit with an error code, not much more we can do at such an  early stage.

**Comment by lyallp** [comment]
After rebuilding gerbera from source, updating the Gentoo startup script /etc/init.d/gerbera to change to a writable directory in the pre-startup functiom, enabling cores in the kernel via /etc/security/limits.conf, rebooting a few times, here is the gdb output of gerbera stack trace.
It's been years since I used gdb in anger, and never on c++, so if you want me to execute some commands whilst in gdb, let me know what they are.
```
# gdb /usr/bin/gerbera /data/cores/core 
GNU gdb (Gentoo 10.1 vanilla) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/gerbera...
(No debugging symbols found in /usr/bin/gerbera)
[New LWP 2437]
[New LWP 2540]
[New LWP 2543]
[New LWP 2538]
[New LWP 2542]
[New LWP 2541]
[New LWP 2539]
[New LWP 2544]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/usr/bin/gerbera --logfile /var/log/gerbera.log --config /etc/gerbera/config.xm'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000055e64d2c740a in std::vector<std::shared_ptr<AutoscanDirectory>, std::allocator<std::shared_ptr<AutoscanDirectory> > >::size() const ()
[Current thread is 1 (Thread 0x7fc81e7acd80 (LWP 2437))]
(gdb) where
#0  0x000055e64d2c740a in std::vector<std::shared_ptr<AutoscanDirectory>, std::allocator<std::shared_ptr<AutoscanDirectory> > >::size() const ()
#1  0x000055e64d2c5fa8 in AutoscanList::size() const ()
#2  0x000055e64d2ddb2c in ContentManager::shutdown() ()
#3  0x000055e64d32d589 in Server::shutdown() ()
#4  0x000055e64d259ba9 in main ()
(gdb) 
```

**Comment by KarlStraussberger** [comment]
Looks like an uninitialized property because of the pending network, maybe database.

**Comment by KarlStraussberger** [comment]
It may have been closed with https://github.com/pupnp/pupnp/pull/235
